File: sol/2140/2140.cpp
struct Solution {
    using ll = long long;
    long long mostPoints(vector<vector<int>>& questions) {
        int n = questions.size();
        vector<ll> memo(n, 0);
        memo[n - 1] = questions[n - 1][0];
        for (int i = n - 2; i >= 0; --i){
            int j = i + questions[i][1] + 1;
            memo[i] = max(memo[i + 1], questions[i][0] + (j < n ? memo[j] : 0));
        }
        return memo[0];
    }
};

File: sol/2140/2140.py
class Solution:
    def mostPoints(self, questions: List[List[int]]) -> int:
        n = len(questions)
        memo = [0] * n
        memo[n - 1] = questions[n - 1][0]
        for i in range(n - 2, -1, -1):
            if (j := i + questions[i][1] + 1) < n:
                memo[i] = max(memo[i + 1], questions[i][0] + memo[j])
            else:
                memo[i] = max(memo[i + 1], questions[i][0])
        return memo[0]
